packages/assets-controllers/src/TokenDetectionController.test.ts

```diff
@@ -438,168 +500,116 @@ describe('TokenDetectionController', () => {
+      let preferencesStateChangeListener: (state: any) => void;
+      const onPreferencesStateChange = sinon.stub().callsFake((listener) => {
+        preferencesStateChangeListener = listener;
+      });
 
-  it('should call getBalancesInSingleCall if onPreferencesStateChange is called with useTokenDetection being true and is changed', async () => {
-    const stub = sinon.stub();
-    const getBalancesInSingleCallMock = sinon.stub();
-    // TODO: Replace `any` with type
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    let preferencesStateChangeListener: (state: any) => void;
-    const onPreferencesStateChange = sinon.stub().callsFake((listener) => {
-      preferencesStateChangeListener = listener;
-    });
-    tokenDetection = new TokenDetectionController(
-      {
+      tokenDetection = new TokenDetectionController({
+        networkClientId: NetworkType.mainnet,
+        selectedAddress: '0x1',
         onPreferencesStateChange,
-        onTokenListStateChange: stub,
-        onNetworkStateChange: stub,
         getBalancesInSingleCall: getBalancesInSingleCallMock,
-        addDetectedTokens: stub,
-        getTokensState: () => tokensController.state,
-        getTokenListState: () => tokenList.state,
-        getNetworkState: () => defaultNetworkState,
         getPreferencesState: () => preferences.state,
-        getNetworkClientById: jest.fn(),
-      },
-      {
-        disabled: false,
-        isDetectionEnabledForNetwork: true,
-        isDetectionEnabledFromPreferences: false,
-        selectedAddress: '0x1',
-      },
-    );
+        messenger: buildTokenDetectionControllerMessenger(controllerMessenger),
+      });
 
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    await preferencesStateChangeListener!({
-      selectedAddress: '0x1',
-      useTokenDetection: true,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      preferencesStateChangeListener!({
+        selectedAddress: '0x2',
+        useTokenDetection: true,
+      });
+      expect(getBalancesInSingleCallMock.calledOnce).toBe(true);
     });
 
-    expect(getBalancesInSingleCallMock.called).toBe(true);
-  });
-
-  it('should call getBalancesInSingleCall if onNetworkStateChange is called with a chainId that supports token detection and is changed', async () => {
-    const stub = sinon.stub();
-    const getBalancesInSingleCallMock = sinon.stub();
-    // TODO: Replace `any` with type
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    let networkStateChangeListener: (state: any) => void;
-    const onNetworkStateChange = sinon.stub().callsFake((listener) => {
-      networkStateChangeListener = listener;
-    });
-    tokenDetection = new TokenDetectionController(
-      {
-        onNetworkStateChange,
-        onTokenListStateChange: stub,
+    it('should be called if network is changed to a chainId that supports token detection', async () => {
```

Comment:
The reason why this test is failing is that the chain ID isn't changing. Even though a `networkClientId` of "polygon" is supplied here, the `getNetworkClientById` action is still mocked to return `mockMainnetClient`, which has a chain ID of 0x1. I think we'll want to make a `mockPolygonClient` and mock `getNetworkClientById` to return that here before calling `changeNetwork`.

Reply:
Looks like this was also a result of misunderstanding this suggestion: https://github.com/MetaMask/core/pull/3609#discussion_r1424681019. I had some of the necessary the unregister/register calls in place here (a26d568a516cc367c15c8f82ccf56d074aa56e4d), but then I removed them and used `.mockReturnValue()` instead. That was causing the failures along with not mocking the polygon client. 

Fixed here: 5a94e14cf51472d39be5a3f3a0fc08f989a5af5e. All tests are finally passing for token detection controller!

Given the current code, is there a way to replace the unregister/register calls with `mockReturnValue` or something similar?

Reply:
Ah, okay. Sorry for the confusion! I thought it would have been sufficient just to use `mockReturnValue`, but perhaps not. I think the way you have it is fine, and we can refactor this later.